### DISCUSSION TITLE
What is the correct way to access information about dual rays?

### DISCUSSION THREAD

**Discussion by febattista** [coin-or/Clp#281] [Q&A] [answered]
Hello everyone,

I'm developing a functionality in another COIN-OR project, i.e., `SYMPHONY` in which I need to collect the following information about a LP Dual solved in `Clp`, that has been proven to be unbounded:

-  A Dual Ray that is the proof of unboundedness; and
- The last Dual Solution just prior the discovery of the Dual Ray;

My final goal is to have a dual solution, say `\pi`, and an extreme ray, say `\sigma`, such that for any scalar `\lambda >= 0` I have that `\pi + \lambda * \sigma` is still dual feasible.

`SYMPHONY` interfaces to `Clp` via `OsiClpSolverInterface` and is build with `Clp master`.

Currently, I've tried to access the aforementioned information by the following functions available in `OsiClpSolverInterface.cpp`

```
std::vector< double * > OsiClpSolverInterface::getDualRays(int /*maxNumRays*/,
  bool fullRay) const
{
  return std::vector< double * >(1, modelPtr_->infeasibilityRay(fullRay));
}
 ```

for the Dual Ray; and 

```
const double * OsiClpSolverInterface::getRowPrice()
```

for the Dual Solution. The problem is that if I combine the results of these functions as mentioned before, the resulting dual solution will not be feasible. Is there anything I'm doing wrong?

Thanks in advance :)

**Comment by jjhforrest** [comment]
To get the duals for the previous iteration, I suggest -
1) add an event handler - there are examples to add an event handler in Cbc, but for Clp look at unitTest.cpp
2) in event handler test for endOfIteration and save sequenceIn, sequenceOut, pivotRow and direction.
3) when unbounded modify basis using that information
4) get duals

**Comment by jjhforrest** [comment, marked as answer]
Added an example event1.cpp to Clp/examples in master.  Done rather too quickly, but it should give you the idea.

**Reply by febattista** [reply]
That's so kind of you, thanks!
